### PR TITLE
Simplify Passes (5/5): Reimplement Simplify as a fixed-point pipeline

### PR DESCRIPTION
**Pull request by tbennun**
- [x] Make simplify a fixed-point pipeline
- [x] Ensure scalability of approach
- [x] Update docs

### CODE REVIEW

File: dace/transformation/passes/fusion_inline.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+@dataclass(unsafe_hash=True)
+class FuseStates(ppl.Pass):
+    """
+    Fuses all possible states of an SDFG (and all sub-SDFGs).
+    """
+    permissive: bool = False  #: If True, ignores some race conditions checks
+    progress: Optional[bool] = None  #: Whether to print progress, or None for default (print after 5 seconds)
+
+    def should_reapply(self, modified: ppl.Modifies) -> bool:
+        return modified & (ppl.Modifies.States | ppl.Modifies.InterstateEdges)
```
Comment (lukastruemper) [inline dace/transformation/passes/fusion_inline.py:23]: Seems like we don't always have a separate test for each pass? As far as I understand, the whole correctness of the compiler now depends on the developer specifying those flags correctly.

Maybe we should stress test each pass in a shotgun approach: Add a test case where we perform randomly generated pipelines on some non-standard SDFGs? Especially the order would be interesting to change

File: doc/source/sdfg/simplify.rst [hunk 2]
```diff
@@ -0,0 +1,68 @@
+.. _simplify:
+
+Simplify Pipeline
```
Comment (lukastruemper) [inline doc/source/sdfg/simplify.rst:3]: Very nice!

File: tests/passes/array_elimination_test.py [hunk 3]
```diff
@@ -1,10 +1,11 @@
 # Copyright 2019-2022 ETH Zurich and the DaCe authors. All rights reserved.
 
 import pytest
+
 import dace
 from dace.sdfg import utils as sdutil
-from dace.transformation.passes.array_elimination import ArrayElimination
 from dace.transformation.pass_pipeline import Pipeline
+from dace.transformation.passes.array_elimination import ArrayElimination
```
Comment (lukastruemper) [inline tests/passes/array_elimination_test.py:8]: Would be easier to review if the tests folder structure would resemble the structure of the dace folder

**Review by lukastruemper — APPROVED** [review]
Goals of the PR:

- Docs: I really appreciate this, awesome work!
- Scalability: I don't see scalability issues now. You also provided empirical numbers on the performance which look very good.

However, I see potential correctness issues in should_reapply. Even if we provide tests for all passes, this is not safe. We should add randomized testing schemes for the simplify pipeline and for different passes.

Another point is the dict with results of previous passes: Maybe this should be wrapped into a data structure that is read-only for other passes. This currently does not protect one faulty pass from messing up other results. Unlikely though

Both points are unrelated from this PR, so I will approve.

It would be nice if the test folder structure had the same structure as the dace package (and matching file names). It is super hard to find the corresponding tests / check if it exists.

**Review by phschaad — CHANGES_REQUESTED** [review]
Very nice changes, thank you for the detailed documentation! Generally LGTM, but 3 smaller questions that I would like answered/addressed quickly.

File: tests/npbench/floyd_warshall_test.py [hunk 4]
```diff
@@ -71,7 +71,7 @@ def run_floyd_warshall(device_type: dace.dtypes.DeviceType):
                                                              'storage': dace.StorageType.FPGA_Local
                                                          }],
                                                          print_report=True)
-        assert sm_applied == 1
+        # assert sm_applied == 1
```
[the remark this answers is not in the document]
Comment (tbennun) [reply tests/npbench/floyd_warshall_test.py:74]: This was commented out after a discussion with Tiziano - based on what happens to the graph (how edge consolidation is configured), different number of StreamingMemory transformations would apply. We ended up saying we do not care how many apply.

[the remark this answers is not in the document]
Comment (tbennun) [reply tests/npbench/floyd_warshall_test.py:74]: @TizianoDeMatteis if you want to chime in

File: tests/persistent_fusion_cudatest.py [hunk 5]
```diff
@@ -4,6 +4,7 @@
 import networkx as nx
 import dace
 from dace.sdfg.graph import SubgraphView
+from dace.sdfg.utils import fuse_states
```
[the remark this answers is not in the document]
Comment (tbennun) [reply tests/persistent_fusion_cudatest.py:7]: Oops, fixing.

**Comment by lukastruemper** [comment]
> 
> I'm not sure what you mean by the unsafe part. This is up to the pass author to define when they think a pass should be rerun. That should be safe regardless if the pass applied again (which should probably just do nothing) or not applied at all (leaving the graph in a slightly less simplified form).
> 

Oh yes, I was referring to the modifies flags and not the rerun function ...

File: tests/npbench/floyd_warshall_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TizianoDeMatteis) [reply tests/npbench/floyd_warshall_test.py:74]: Yep, I agree!

[the remark this answers is not in the document]
Comment (tbennun) [reply tests/npbench/floyd_warshall_test.py]: I removed it now.

[the remark this answers is not in the document]
Comment (phschaad) [reply tests/npbench/floyd_warshall_test.py]: Ok makes sense, thank you for the clarification :-)

**Pull request merged**